### PR TITLE
TraceKit can now be used on frames

### PR DESCRIPTION
**Pull request by aquadk**
Hi, have made this change so, I only need to include TraceKit once on my frame page. 

Here is the code for the frame page.

    function insertTraceKit() {

      function yourLogger(errorReport) {
        console.log(errorReport);
        return true;
      }

      TraceKit.report.subscribe(yourLogger);

      function wrapped(aThis) {
        TraceKit.report.subscribe(yourLogger, aThis.target.contentWindow);
        try {
          aThis.target.contentWindow.onunload = unwrapped;
        } catch (e) {
          //console.log('cross-origin frame detected');
        }
      }

      function unwrapped(aThis) {
        TraceKit.report.unsubscribe(yourLogger, aThis.currentTarget);
      }

      var frames = document.getElementsByTagName('frame');
      for (var i = 0; i < frames.length; i++) {
        frames[i].onload = wrapped;
      };

    }

    var readyStateCheckInterval = setInterval(waitInit, 10, this);

    function waitInit() {
      //console.log(document.readyState);
      if (document.readyState === "interactive") {
        clearInterval(readyStateCheckInterval);
        insertTraceKit();
      }
    }

### CODE REVIEW

File: tracekit.js [hunk 1]
```diff
@@ -921,7 +935,7 @@ TraceKit.computeStackTrace = (function computeStackTraceWrapper() {
                     }
                 }
             } else if ((parts = lineRE3.exec(lines[line]))) {
-                var url = window.location.href.replace(/#.*$/, '');
+          var url = TraceKit.windowPointer.location.href.replace(/#.*$/, '');
```
Comment (niemyjski) [inline tracekit.js]: What about parsing source in the inner frame?

File: tracekit.js [hunk 2]
```diff
@@ -880,7 +894,7 @@ TraceKit.computeStackTrace = (function computeStackTraceWrapper() {
             lineRE2 = /^\s*Line (\d+) of inline#(\d+) script in ((?:file|https?|blob)\S+)(?:: in function (\S+))?\s*$/i,
             lineRE3 = /^\s*Line (\d+) of function script\s*$/i,
             stack = [],
-            scripts = (window && window.document && window.document.getElementsByTagName('script')),
+        scripts = (TraceKit.windowPointer && TraceKit.windowPointer.document && TraceKit.windowPointer.document.getElementsByTagName('script')),
```
Comment (niemyjski) [inline tracekit.js]: Can you move this window pointer to a local variable, this will not minify very well.

File: tracekit.js [hunk 3]
```diff
@@ -419,7 +431,9 @@ TraceKit.computeStackTrace = (function computeStackTraceWrapper() {
             */
             var source = '';
             var domain = '';
-            try { domain = window.document.domain; } catch (e) { }
+        try {
+          domain = window.document.domain;
```
Comment (niemyjski) [inline tracekit.js]: Why not use the pointer here>

File: tracekit.js [hunk 4]
```diff
@@ -218,13 +229,14 @@ TraceKit.report = (function reportModuleWrapper() {
...
-    function installGlobalHandler () {
-        if (_onErrorHandlerInstalled === true) {
+    function installGlobalHandler(aHandlers, aWindow) {
+      if (aWindow._onErrorHandlerInstalled === true) {
             return;
         }
-        _oldOnerrorHandler = window.onerror;
-        window.onerror = traceKitWindowOnError;
-        _onErrorHandlerInstalled = true;
+      var _oldOnerrorHandler = aWindow.onerror;
```
Comment (niemyjski) [inline tracekit.js]: if this is no longer a private field we need to change the variable name

File: tracekit.js [hunk 5]
```diff
@@ -204,12 +217,10 @@ TraceKit.report = (function reportModuleWrapper() {
               'stack': [location]
             };
 
-            notifyHandlers(stack, true, null);
+        notifyHandlers(stack, true, arguments);
         }
 
-        if (_oldOnerrorHandler) {
```
Comment (niemyjski) [inline tracekit.js]: Any reason for removing this?

File: tracekit.js [hunk 6]
```diff
@@ -149,27 +160,29 @@ TraceKit.report = (function reportModuleWrapper() {
...
         var exception = null;
         if (isWindowError && !TraceKit.collectWindowErrors) {
           return;
         }
         for (var i in handlers) {
-            if (_has(handlers, i)) {
+        if (_has(handlers, i) && handlers[i][1] === TraceKit.windowPointer) {
                 try {
-                    handlers[i](stack, isWindowError, error);
+            handlers[i][0](stack, isWindowError, (aArguments.length > 4) ? aArguments[4] : null);
```
Comment (niemyjski) [inline tracekit.js]: What's the 3rd and 4th parameter here? This became much less readable.

File: tracekit.js [hunk 7]
```diff
@@ -149,27 +160,29 @@ TraceKit.report = (function reportModuleWrapper() {
...
                 } catch (inner) {
                     exception = inner;
                 }
+          if (handlers[i][1]._oldOnerrorHandler) {
```
Comment (niemyjski) [inline tracekit.js]: I thought this was a local variable.

**Comment by aquadk** [comment]
Hi,
Have now made the changes you asked for.

File: tracekit.js (hunk 5 above)
Comment (aquadk) [inline tracekit.js]: is moved to notifyHandlers

File: tracekit.js [hunk 8]
```diff
@@ -149,27 +160,31 @@ TraceKit.report = (function reportModuleWrapper() {
      * @memberof TraceKit.report
      * @throws An exception if an error occurs while calling an handler.
      */
-    function notifyHandlers(stack, isWindowError, error) {
+    function notifyHandlers(stack, isWindowError, aArguments) {
```
Comment (niemyjski) [inline tracekit.js]: What does arguments contain? Can we call this args? and maybe a comment to break down what the possible structure is.

File: tracekit.js [hunk 9]
```diff
@@ -149,27 +160,31 @@ TraceKit.report = (function reportModuleWrapper() {
...
         if (isWindowError && !TraceKit.collectWindowErrors) {
           return;
         }
         for (var i in handlers) {
-            if (_has(handlers, i)) {
+        if (_has(handlers, i) && handlers[i][1] === TraceKit.windowPointer) {
                 try {
-                    handlers[i](stack, isWindowError, error);
+                  var errorObj=(aArguments.length > 4) ? aArguments[4] : null;
+                  handlers[i][0](stack, isWindowError, errorObj);
```
Comment (niemyjski) [inline tracekit.js]: This is kind of confusing, so we call a method with stack, isWindowErorr and an array with 4 items? We seem to only be using the error object, so it might be better to revert this signature to just be error object.  Previous thought: Maybe we need to add a comment stating that these arguments are actually: `message, url, lineNo, columnNo, errorObj` (if we were on es6, restructuring would make this so much more readable but we can't use that :(`,

File: tracekit.js [hunk 10]
```diff
@@ -149,27 +160,31 @@ TraceKit.report = (function reportModuleWrapper() {
...
                 } catch (inner) {
                     exception = inner;
                 }
+          // Call old onerror events
+          if (handlers[i][2]) {
```
Comment (niemyjski) [inline tracekit.js]: We weren't doing this before were we? It's just really hard to follow this and make sense of what's going on..  We call all the other handlers in a try catch.

File: tracekit.js [hunk 11]
```diff
@@ -183,13 +198,13 @@ TraceKit.report = (function reportModuleWrapper() {
      */
     function traceKitWindowOnError(message, url, lineNo, columnNo, errorObj) {
         var stack = null;
-
+      TraceKit.windowPointer = this;
```
Comment (niemyjski) [inline tracekit.js]: So this modifies the pointer which might of been set earlier when you subscribed. I'm not sure this should be updated as it would change the subscription handler? Thoughts?

File: tracekit.js [hunk 12]
```diff
@@ -118,27 +118,38 @@ TraceKit.report = (function reportModuleWrapper() {
...
      * @param {Function} handler
      * @memberof TraceKit.report
      */
-    function unsubscribe(handler) {
+    function unsubscribe(handler, win) {
+      win = (win || window);
+      if (isWindowAccessible(win)) {
         for (var i = handlers.length - 1; i >= 0; --i) {
-            if (handlers[i] === handler) {
+          if (handlers[i][0] === handler && win === handlers[i][1]) {
```
Comment (niemyjski) [inline tracekit.js]: What's all stored on handler now in this array?

File: tracekit.js [hunk 13]
```diff
@@ -361,9 +375,10 @@ TraceKit.report = (function reportModuleWrapper() {
  * @memberof TraceKit
  * @namespace
  */
-TraceKit.computeStackTrace = (function computeStackTraceWrapper() {
+TraceKit.computeStackTrace = (function() {
     var debug = false,
-        sourceCache = {};
+        sourceCache = {},
+        curWin = TraceKit.windowPointer;
```
Comment (niemyjski) [inline tracekit.js]: Can we rename this to win

File: tracekit.js [hunk 14]
```diff
@@ -1109,6 +1126,7 @@ TraceKit.computeStackTrace = (function computeStackTraceWrapper() {
      * @memberof TraceKit.computeStackTrace
      */
     function computeStackTrace(ex, depth) {
+        curWin = TraceKit.windowPointer;
```
Comment (niemyjski) [inline tracekit.js]: This doesn't seem to be used

File: tracekit.js [hunk 15]
```diff
@@ -1195,8 +1213,8 @@ TraceKit.computeStackTrace = (function computeStackTraceWrapper() {
  */
 TraceKit.extendToAsynchronousCallbacks = function () {
     var _helper = function _helper(fnName) {
-        var originalFn = window[fnName];
-        window[fnName] = function traceKitAsyncExtension() {
+      var originalFn = TraceKit.windowPointer[fnName];
+      TraceKit.windowPointer[fnName] = function traceKitAsyncExtension() {
```
Comment (niemyjski) [inline tracekit.js]: Might want to store this in a variable

File: tracekit.js (hunk 12 above)
Comment (niemyjski) [inline tracekit.js]: We weren't doing this before, but since you are calling unsubscribe and you have an array with the previous error handler. Do you think we should be setting it back?

Comment (aquadk) [inline tracekit.js]: Good point, i will insert the code

File: tracekit.js (hunk 8 above)
Comment (aquadk) [inline tracekit.js]: ok, arguments contains all the arguments from onerror event

**Comment by niemyjski** [comment]
I'm sorry for taking so long to get back to this. What is the current status of this?

**Comment by aquadk** [comment]
I have updated the code for Iframe page too.
```

   /**    
     * Description    
     * @param {type} aTracefunc Description    
     */
    function insertTraceKit(aTracefunc) {

      TraceKit.report.subscribe(aTracefunc);

      function wrapped(aThis) {
        var w = (aThis.target || aThis);

        function docUnloadChange(e) {
          TraceKit.report.unsubscribe(yourLogger, e.currentTarget);
          //console.log('unload', e.currentTarget.location.href, e.currentTarget.readyState);
          var l = e.currentTarget.location.href;
          var checkForUnload = function (e) {
            try {
              if (l != e.location.href) {
                clearInterval(unloadCheckInterval);
                //console.log(e.location.href);
                TraceKit.report.subscribe(aTracefunc, e);
              }
            } catch (exception) {
              clearInterval(unloadCheckInterval);
              unloadCheckInterval = setInterval(checkForUnload, 100, this);
              //console.log('cross-origin frame detected');
            }
          };
          var unloadCheckInterval = setInterval(checkForUnload, 1, this);

        }

        TraceKit.report.subscribe(aTracefunc, w.contentWindow);
        try {
          w.contentWindow.addEventListener("unload", docUnloadChange);
        } catch (e) {
          //console.log('cross-origin frame detected');
        }
      }

      var frames = document.querySelectorAll('frame,iframe');
      for (var i = 0; i < frames.length; i++) {
        frames[i].onload = wrapped;
        try {
          // console.log(frames[i].contentDocument.readyState, frames[i].contentDocument.location.href);
          frames[i].readyStateCheckInterval = setInterval(fasteractiveTrace, 1, frames[i]);
        } catch (e) {
          //console.log('cross-origin frame detected');
        }
      }

      function fasteractiveTrace(e) {
        try {
          if (e.contentDocument) {
            if (e.contentDocument.readyState == "complete" || e.contentDocument.readyState == "interactive") {
              wrapped(e);
              clearInterval(e.readyStateCheckInterval);
            }
            // console.log(e.contentDocument.readyState, e.contentDocument.location.href);
          }
        } catch (exception) {
          //console.log('cross-origin frame detected');
        }
      }

    }

    var readyStateCheckInterval = setInterval(waitInit, 10, this);

    function waitInit() {
      //console.log(document.readyState);
      if (document.readyState === "complete" || document.readyState === "interactive") {
        clearInterval(readyStateCheckInterval);
        insertTraceKit(yourLogger);
      }
    }
```

**Comment by niemyjski** [comment]
Thanks, were any other changes required for this pr? Do you think we should put this in a wiki entry?

**Comment by niemyjski** [comment]
Thanks, were any other changes required for this pr? Do you think we should put this in a wiki entry?

**Comment by niemyjski** [comment]
Thanks, were any other changes required for this pr? Do you think we should put this in a wiki entry? I'd like to get this merged in and cleaned up

File: tracekit.js [hunk 16]
```diff
@@ -291,8 +302,8 @@ TraceKit.report = (function reportModuleWrapper() {
         // slow slow IE to see if onerror occurs or not before reporting
         // this exception; otherwise, we will end up with an incomplete
         // stack trace
-        setTimeout(function () {
-            if (lastException === ex) {
+        window.setTimeout(function () {
```
Comment (niemyjski) [inline tracekit.js]: Any specific reason to putting this to window? It might block users from using this in node?

File: tracekit.js [hunk 17]
```diff
@@ -224,17 +234,16 @@ TraceKit.report = (function reportModuleWrapper() {
...
                 'name': name,
                 'message': msg,
-                'mode': 'onerror',
-                'stack': [location]
+              'mode': 'onerror',
+              'stack': [location]
             };
 
-        notifyHandlers(stack, true, arguments);
+            notifyHandlers(stack, true, null);
```
Comment (niemyjski) [inline tracekit.js]: any reason for not passing through the arguments?

File: tracekit.js [hunk 18]
```diff
@@ -204,17 +215,16 @@ TraceKit.report = (function reportModuleWrapper() {
       TraceKit.windowPointer = this;
         if (lastExceptionStack) {
             TraceKit.computeStackTrace.augmentStackTraceWithInitialElement(lastExceptionStack, url, lineNo, message);
-        processLastException(arguments);
+    	    processLastException();
```
Comment (niemyjski) [inline tracekit.js]: any reason for not passing through the arguments?

**Comment by aquadk** [comment]
I have merged our code and committed all changes.

**Comment by aquadk** [comment]
Forgot to push all changes.. sorry

**Comment by niemyjski** [comment]
Hello, I'm not seeing any of the latest changes, can you please review the files changed here as you can see we added more code around installing additional global handlers.

**Comment by aquadk** [comment]
All code are pushed into aquadk/Tracekit master branch. Including global handlers.

File: tracekit.js [hunk 19]
```diff
@@ -216,22 +245,21 @@ TraceKit.report = (function reportModuleWrapper() {
...
+                incomplete: false,
+                partial: false
             };
 
-            notifyHandlers(stack, true, null);
+            notifyHandlers(stack, true, Array.prototype.slice.call(arguments));
         }
 
-        if (_oldOnerrorHandler) {
-            return _oldOnerrorHandler.apply(this, arguments);
```
Comment (niemyjski) [inline tracekit.js]: I think we were calling old handlers to preserve existing work flows.

File: tsconfig.json [hunk 20]
```diff
@@ -0,0 +1,12 @@
+{
+    "compilerOptions": {
+        "module": "commonjs",
+        "target": "es6",
```
Comment (niemyjski) [inline tsconfig.json]: We actually have to target es3 everywhere :\.

**Comment by niemyjski** [comment]
Looks good, I added two comments on things, if we feel we are good I'll merge it in.

**Comment by niemyjski** [comment]
@aquadk Could you please review the feedback so we can take your great work and get it merged in.

File: tracekit.d.ts [hunk 21]
```diff
@@ -7,6 +7,22 @@ export interface StackFrame {
     context:string[];
 }
 
+declare global {
+    interface Window {
+        onunhandledrejection: PromiseRejectionEvent;
+        _onErrorHandlerInstalled : boolean;
```
[the remark this answers is not in the document]
Comment (niemyjski) [reply tracekit.d.ts]: I don't think so.

**Comment by niemyjski** [comment]
Can we get this updated so we can get it merged in.

**Comment by niemyjski** [comment]
Can we get this updated so we can get it merged in.

**Comment by niemyjski** [comment]
Can we get this updated so we can get it merged in.
